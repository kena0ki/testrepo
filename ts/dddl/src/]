/** Options for data to be generated. */
export type GeneratorOptions = {
  /** Output format. Either csv or insert statement. */
  outputFormat: CsvFormat|InsertStatementFormat,
  /**
   *  Options for each column to determine how they are generated.
   *  The options need to be valid types for corresponding columns.
   *    - Column -                   - Option -
   *    Numeric                   -> NumericColumnOption
   *    Normal character          -> CharColumnOption
   *    Binary character          -> BinaryColumnOption (not yet implemented)
   *    Date, Time, Timestamp     -> DateColumnOption (not yet implemented)
   *  If you want a fixed value for some columns throughout entire data,
   *  you can specify the value using FixedValue.
   */
  columnOptions?: IntegerColumnOption[],
  /**
   *  
   */
  columnOptionsDefault?: IntegerColumnOption[],
  /** A function to manipulate row data after it's generated. */
  eachRow?: (rowCount: number, columns: string[], prev: object) => string,
  /** The number of rows to be generated. */
  rows: number,
  /**
   * How to behave when invalid data is detected. Default: log.
   *   log: continue, but output log.
   *   abort: abort data generation.
   *   ignore: ignore error data and going on.
   */
  errorAction: 'log'|'abort'|'ignore',
}
/** CsvFormat is used for GeneratorOptions.outputFormat */
export class CsvFormat {
  /** Delimiter of each column. Default: ',' */
  public delimiter?: string = `,`
  /** Quote for each value. Default: '"' */
  public quote?: string = `"`
  /** Escape sequence. Default: '"' */
  public escapeSequence?: string = `"`
  /** Define options */
  public define(obj: OnlyOfType<string|number,CsvFormat>): this {
    (Object.keys(obj) as [keyof CsvFormat]).forEach(key => {
      if (isOfTypeNumber<CsvFormat>(key)) this[key] = obj[key];
      if (isOfTypeString<CsvFormat>(key)) this[key] = obj[key];
    });
    return this;
  }
}
/** InsertStatementFormat is used for GeneratorOptions.outputFormat */
export class InsertStatementFormat {
}
/** NumericColumnOptions is used for GeneratorOptions.columnOptions */
export class IntegerColumnOption {
  /** How much advance per each row. Default: 1 for integers, 0.1 for decimals */
  public stepBy?: number
  /** Value of the first row */
  public initialValue?: number = 1
  /** Prefix. Default: nothing */
  public prefix: Prefix = ''
  /** Limit of incrementation. Default: depend on the corresponding column data type. */
  public limit?: number = Infinity
  /**
   * How to behave when incrementation hits the limit. Default: loop.
   *   loop: back to the initial value and continue to increment
   *   negate: negate the value and continue to increment
   *   keep: stop incrementation and keep the limit value
   */
  public loop?: 'loop'|'negate'|'keep' = 'loop'
}
/** CharColumnOptions is used for GeneratorOptions.columnOptions */
export class CharColumnOption {
  /** Limit of incrementation. Default: depend on the corresponding column data type. */
  public maxLength: number = Infinity
  /** Prefix. Default: column number */
  public prefix: Prefix = ''
  /** Suffix. Default: nothing */
  public suffix: Prefix = ''
  /**
   * How to behave when incrementation hits the limit. Default: loop.
   *   loop: back to the initial value and continue to increment
   *   keep: stop incrementation and keep the limit value
   */
  public loop?: 'loop'|'keep' = 'loop'
}
/** BinaryColumnOptions is used for GeneratorOptions.columnOptions */
export class BinaryColumnOptions {
  // TODO
}
/** DateColumnOptions is used for GeneratorOptions.columnOptions */
export class DateColumnOptions {
  // TODO
}
/** FixedValue is use for GeneratorOptions.columnOptions */
export class FixedValue {
  constructor(
    /** The value to be set to a column */
    public value: string = '',
  ) {}
}
type OnlyKeysOfType<T,O> = Extract<keyof O, {[K in keyof O]: O[K] extends T|undefined ? K : never}[keyof O]>
type OnlyOfType<T,O> = {[K in OnlyKeysOfType<T,O>]?: O[K]};
function isOfTypeNumber<K>(key: keyof K): key is OnlyKeysOfType<number,K> { return typeof key === 'number'; }
function isOfTypeString<K>(key: keyof K): key is OnlyKeysOfType<string,K> { return typeof key === 'string'; }
type Prefix = string|((row: number, col: number, colName: string) => string);

/**
 * Generates data from a create table statement with options.
 * How data is generated depends on types and options but the general idea is that
 * generator adds 1 to previous data row by row so each column would have sequentially incremented number.
 * Given that we have following ddl, then we get following data.
 */
export const generate = (ddl: string, options: GeneratorOptions) => {
};

