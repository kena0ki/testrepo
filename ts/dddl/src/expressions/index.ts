export * as exprs from './expressions';
export * from './expressions';
export * as values from './values';
export * from './values';
export * as ops from './operators';

