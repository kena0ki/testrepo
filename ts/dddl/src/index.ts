export * as parser from './parser';
export * from './parser';
export * as generator from './generator';
export * from './generator';
export * as dataTypes from './data-types';
// export * from './data-types';

