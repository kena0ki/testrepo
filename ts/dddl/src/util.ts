// import * as util from 'util';

export const global = Function('return this')();

export const logger = global.dddlEnableLog || (global.process && global.process.env.DDDL_ENABLE_LOG) ? console : {
  log(){},
  trace(){},
  time(){},
  timeEnd(){},
};
// export const logger = console;

// export function wrapError<T extends Error>(wrapper: T, cause: Error): T {
//   wrapper.message = wrapper.message + '\n' + cause.message;
//   if (wrapper.stack) {
//     wrapper.stack = wrapper.stack.split('\n').slice(0,2).join('\n') + '\n' + cause.stack;
//   }
//   return wrapper;
// }

export const min = (...args: number[]): number => args.sort((a,b) => a-b)[0];
export const max = (...args: number[]): number => args.sort((a,b) => b-a)[0];

const surrogatePairs = /[\uD800-\uDBFF][\uDC00-\uDFFF]/g;
export const length = (str: string): number => str.replace(surrogatePairs,' ').length;

// dirty hack for calculation of floating point number.
// TODO overflow
export const add = (left:number, right:number, scale:number): number => {
  if (scale <= 0) return (+left)+(+right);
  const digits=10**scale;
  return ((left*digits)+(right*digits))/digits;
};
export const subtract = (left:number, right:number, scale:number): number => {
  if (scale <= 0) return left-right;
  const digits=10**scale;
  return ((left*digits)-(right*digits))/digits;
};

export * from './util';
export * as util from './util';

