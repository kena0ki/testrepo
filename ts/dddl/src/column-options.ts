import { ObjectName, Ident, ReferencialAction } from './parser';
import { Expr } from './expressions';

export class ColumnOption {
  private columnoption: never
  constructor() {}
}
export class NotNull extends ColumnOption {}
export class Null extends ColumnOption {}
export class Default extends ColumnOption {
  constructor(public expr: Expr) { super(); }
}
export class Unique extends ColumnOption {
  constructor(public isPrimary: boolean) { super(); }
}
export class Foreign extends ColumnOption {
  constructor(
    public foreignTable: ObjectName,
    public referredColumns: Ident[],
    public onDelete?: ReferencialAction,
    public onUpdate?: ReferencialAction,
  ) { super(); }
}
export class Check extends ColumnOption {
  constructor(public expr: Expr) { super(); }
}
export class DiarectSpecific extends ColumnOption {
  constructor(public name: 'AUTO_INCREMENT'|'AUTOINCREMENT') { super(); }
}

export * from './column-options';
export * as columnOptions from './column-options';

