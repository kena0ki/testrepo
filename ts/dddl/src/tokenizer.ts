import { logger, wrapError } from './util';
const KEYWORDS = [ // only contains keywords that is necessary for create statements
  'CREATE',
  'TABLE',
  'OR',
  'REPLACE',
  'IF',
  'NOT',
  'EXISTS',
] as const;
type Keyword = typeof KEYWORDS[number];
export class Token {
  constructor(private _value: string) { }
  get value(): string { return this._value; }
  get length(): number { return this._value.length; }
  get class(): string { return this.constructor.name; }
}
const equalToKeyword = (token: Token, keyword: Keyword): boolean => token instanceof Word && token.value === keyword;
export const tokenUtil = {
  equalToKeyword,
};
export class Whitespace extends Token {}
export class NewLine extends Whitespace {}
export class Word extends Token {} // keyword or identifier
export class DelimitedIdent extends Token {
  constructor(private _content: string, private _delimiter: string) { super(_delimiter + _content + _delimiter); }
  get content(): string { return this._content; }
  get delimiter(): string { return this._delimiter; }
}
export class SingleQuotedString extends Token {}
export class Num extends Token {}
export class LParen extends Token {}
export class RParen extends Token {}
export class Comma extends Token {}
export class Cmmnt extends Token {}
export class Operator extends Token {}
export class Other extends Token {}

export class NonCharcterStringLiteral extends Token {
  constructor(private _content: string, private _prefix: string) { super(_prefix + _content); }
  get content(): string { return this._content; }
  get prefix(): string { return this._prefix; }
}
export class NationalStringLiteral extends NonCharcterStringLiteral {}
export class HexStringLiteral extends NonCharcterStringLiteral {}

export const SPACE = new Whitespace(' ');
export const TAB = new Whitespace('\t');
export const LF = new NewLine('\n');
export const CRLF = new NewLine('\r\n');
export const CR = new NewLine('\r');
export const LPAREN = new LParen('(');
export const RPAREN = new RParen(')');
export const COMMA = new Comma(',');
export const CONCAT = new Operator('||');
export const EQ = new Operator('=');
export const RARROW = new Operator('=>');
export const PERIOD = new Operator('.');
export const PLUS = new Operator('+');
export const MINUS = new Operator('-');
export const DIV = new Operator('/');
export const MULT = new Operator('*');
export const MOD = new Operator('%');
export const NEQ = new Operator('!=');
export const DOUBLE_EXCLAMATION_MARK = new Operator('!!');
export const EXCLAMATION_MARK = new Operator('!');
export const LT = new Operator('<');
export const LTEQ = new Operator('<=');
export const NEQ2 = new Operator('<>');
export const SHIFT_LEFT = new Operator('<<');
export const GT = new Operator('>');
export const GTEQ = new Operator('>=');
export const SHIFT_RIGHT = new Operator('>>');
export const COLON = new Operator(':');
export const DOUBLE_COLON = new Operator('::');
export const SEMICOLON = new Operator(';');
export const BACKSLASH = new Operator('\\');
export const LBRACKET = new Operator('[');
export const RBRACKET = new Operator(']');
export const AMPERSAND = new Operator('&');
export const CARET = new Operator('^');
export const LBRACE = new Operator('{');
export const RBRACE = new Operator('}');
export const TILDE = new Operator('~');
export const HASH = new Operator('#');
export const ATSIGN = new Operator('@');

const TOKENIZE_SINGLE_QUOTED_STRING_ERROR = 'Unterminated string literal';
const TOKENIZE_DELIMITED_STRING_ERROR = (delimiter => `Expected close delimiter ${delimiter} before EOF.`)('"');
const TOKENIZE_MULTI_LINE_COMMENT_ERROR = 'Unexpected EOF while in a multi-line comment';


type Rule = { [ch: string]: (chars: string[]) => Token };
const rule: Rule = {
  // whitespaces
  ' ': () => SPACE,
  '\t': () => TAB,
  '\n': () => LF,
  '\r': chars => chars[1] === '\n' ? CRLF : CR,
  // national string: N'...'
  'N': chars => chars[1] === '\'' ? new NationalStringLiteral(tokenizeSingleQuotedString(chars.slice(1)), chars[0]) : new Word(tokenizeIdentifier(chars)),
  // hex string: X'...' or x'...'. the latter (small x) is PostgreSQL diarect.
  'X': chars => chars[1] === '\'' ? new HexStringLiteral(tokenizeSingleQuotedString(chars.slice(1)), chars[0]) : new Word(tokenizeIdentifier(chars)),
  'x': chars => chars[1] === '\'' ? new HexStringLiteral(tokenizeSingleQuotedString(chars.slice(1)), chars[0]) : new Word(tokenizeIdentifier(chars)),
  // identifier or keyword
  ident: chars => new Word(tokenizeIdentifier(chars)),
  // string
  '\'': chars => new SingleQuotedString(tokenizeSingleQuotedString(chars)),
  // delimited (quoted) identifier
  delimitedIdent: chars => new DelimitedIdent(takeWhileOrError(chars.slice(1), 1, c => ~~(c !== chars[0]), TOKENIZE_DELIMITED_STRING_ERROR), chars[0]),
  // number
  num: chars => new Num(takeWhile(chars, 1, c => ~~(isDigit(c) || c === '.')) ),
  // puctuations
  '(': () => LPAREN,
  ')': () => RPAREN,
  ',': () => COMMA,
  // operators
  '-': chars => chars[1] === '-' ? new Cmmnt(takeWhile(chars, 2, c => ~~(c !== '\n')  /* TODO only LF? */ )) : MINUS,
  '/': chars => chars[1] === '*' ? new Cmmnt(takeWhileOrError(chars, 4, (c,i,chars) => ~~!(chars[i-2] === '*' && chars[i-1] === '/'), TOKENIZE_MULTI_LINE_COMMENT_ERROR)) : DIV,
  '+': () => PLUS,
  '*': () => MULT,
  '%': () => MOD,
  '|': chars => chars[1] === '|' ? CONCAT : new Other('|'),
  '=': chars => chars[1] === '>' ? RARROW : EQ,
  '.': () => PERIOD,
  '!': chars => chars[1] === '=' ? NEQ : chars[1] === '!' ? DOUBLE_EXCLAMATION_MARK : EXCLAMATION_MARK,
  '<': chars => chars[1] === '=' ? LTEQ : chars[1] === '>' ? NEQ2 : chars[1] === '<' ? SHIFT_LEFT : LT,
  '>': chars => chars[1] === '=' ? GTEQ : chars[1] === '>' ? SHIFT_RIGHT : GT,
  ':': chars => chars[1] === ':' ? DOUBLE_COLON : COLON,
  ';':  () => SEMICOLON,
  '\\': () => BACKSLASH,
  '[':  () => LBRACKET,
  ']':  () => RBRACKET,
  '&':  () => AMPERSAND,
  '^':  () => CARET,
  '{':  () => LBRACE,
  '}':  () => RBRACE,
  '~':  () => TILDE,
  '#':  () => HASH, // TODO dead code? this char is in isIdentifierStart
  '@':  () => ATSIGN, // TODO dead code? this char is in isIdentifierStart
};

class TokenizeError extends Error {
  constructor(msg: string, public row: number, public col: number){
    super(msg);
  }
}

export class TokenSet extends Array<Token> {
  constructor(...a) { super(...a); }
  get tokens(): Token[] { return this.map(v=>v); }
  toString = () => {
    const tokens = this.slice(1).reduce(
      (prev, curr) => `${prev},\n  ${curr.constructor.name}: ${JSON.stringify(curr)}`,
      `  ${this[0].constructor.name}: ${JSON.stringify(this[0].value)}`
    );
    return `[\n${tokens}\n]`;
  }
  joinValues = (delim:string = ','): string => this.slice(1).reduce((prev, curr) => `${prev}${delim}${curr.value}`, this[0].value);
}

const nextMeaningfulTokenIdx = (tokenSet: TokenSet, start: number): number => {
  let i=start;
  while(i<tokenSet.length && tokenSet[i] instanceof Whitespace) i++;
  return i;
};
const parseKeyword = (tokenSet: TokenSet, start: number, keyword: Keyword): number => {
  return tokenUtil.equalToKeyword(tokenSet[start], keyword) ? nextMeaningfulTokenIdx(tokenSet, start) : -1;
};
const parseKeywords = (tokenSet: TokenSet, start: number, keywords: Keyword[]): number => {
  let i=start;
  for (let j=0; i<tokenSet.length && j<keywords.length; j++) {
    if((i = tokenSetUtil.parseKeyword(tokenSet, start, keywords[j])) < 0) break;
  }
  return i;
};
export const tokenSetUtil = {
  nextMeaningfulTokenIdx,
  parseKeyword,
  parseKeywords,
};

export const tokenize = (src: string): TokenSet => {
  const chars = Array.from(src);
  const tokens: TokenSet = new TokenSet();
  let row = 1;
  let col = 1;
  let i=0;
  while(i < chars.length) {
    const c = chars[i];
    const key = ['N','X','x'].includes(c) ? c :
                isIdentifierStart(c) ? 'ident' :
                isDelimitedIdentifierStart(c) ? 'delimitedIdent' :
                isDigit(c) ? 'num' :
                c;
    try {
      const token = rule[key]?.(chars.slice(i)) || new Other(key);
      logger.log('c:', `|${c}|`);
      logger.log('token:', `*${token.value}*`);
      i += token.length;
      tokens.push(token);
      if (token instanceof NewLine) {
        row++;
        col=1;
      } else {
        col += token.length;
      }
    } catch (err) {
      const e = wrapError(new TokenizeError('Tokenize error', row, col), err);
      logger.log(e.stack);
      throw e;
    }
  }
  return tokens;
};
const min = (...args: number[]): number => args.sort((a,b) => a-b)[0];
const takeWhile = (chars: string[], testStart: number, advanceBy: (ch: string, idx: number, chars: string[]) => number): string => {
  let i=testStart;
  let by=0;
  while(i < chars.length && (by = advanceBy(chars[i], i, chars)) ) i+=by;
  logger.log(chars, i, chars[i]);
  return chars.slice(0, min(i, chars.length)).join('');
};
const takeWhileOrError = (chars: string[], testStart: number, advanceBy: (ch: string, idx: number, chars: string[]) => number, msg: string): string => {
  let i=testStart;
  let by=0;
  while(i < chars.length && (by = advanceBy(chars[i], i, chars)) ) i+=by;
  logger.log(chars, i, chars[i]);
  if (i === chars.length) throw new Error(msg);
  return chars.slice(0, i).join('');
};
const isDigit = (ch: string): boolean => '0' <= ch && ch <= '9';
const delimiters = ['"']; // TODO dialect
const isDelimitedIdentifierStart = (ch: string): boolean => delimiters.includes(ch);
const isIdentifierStart = (ch: string): boolean => ['@', '#', '_'].includes(ch) ||  ('A' <= ch && ch <= 'Z' ) || ('a' <= ch && ch <= 'z');
const isIdentifierPart = (ch: string): boolean => ['@', '$', '#', '_'].includes(ch) || ('0' <= ch && ch <= '9') || ('A' <= ch && ch <= 'Z' ) || ('a' <= ch && ch <= 'z');
const tokenizeSingleQuotedString = (chars: string[]): string => takeWhileOrError(chars, 2, tokenizeSingleQuotedStringAdvanceBy, TOKENIZE_SINGLE_QUOTED_STRING_ERROR);
const tokenizeSingleQuotedStringAdvanceBy = (c:string,i:number,chars:string[]):number => chars[i-1] !== '\'' ? 1 : (chars[i-1] === '\'' && c === '\'') ? 2 : 0;
const tokenizeIdentifier = (chars: string[]): string => takeWhile(chars, 1, c => ~~isIdentifierPart(c));

