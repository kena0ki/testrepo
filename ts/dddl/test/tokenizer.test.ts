import { tokenize  } from '../src/tokenizer';
import { logger } from '../src/util';
import { promises as fs } from 'fs';
import * as path from 'path';

class A extends Object {
  get hoge() { return 1;}
}
console.log('hogeeeee', (new A).hoge);

describe('tokenize', () => {
  test(`001. tokenize create statement`, async () => {
    const file = path.join(__dirname, `sql/tokenizer_001.sql`);
    const sql = await fs.readFile(file, 'utf8');
    const tokenSet = tokenize(sql);
    logger.log(tokenSet.tokens);
    expect(tokenSet).toMatchSnapshot('Create table');
  });
  test(`002. tokenize select statement`, async () => {
    const start = Date.now();
    const file = path.join(__dirname, `sql/tokenizer_002.sql`);
    const sql = await fs.readFile(file, 'utf8');
    const tokenSet = tokenize(sql);
    logger.log('time:', Date.now() - start);
    logger.log(tokenSet.toString());
    logger.log(tokenSet.joinValues(''));
    expect(tokenSet).toMatchSnapshot('query');
  });
  test(`003. tokenize operators`, async () => {
    const tokenSet = tokenize(`\\[]&^{}~#`);
    logger.log(tokenSet);
    expect(tokenSet).toMatchSnapshot('operators');
  });
  test(`004. tokenize error`, async () => {
    try {
      tokenize(`select 'error`);
    } catch (err) {
      logger.log(err);
      expect(err).toMatchSnapshot('error');
    }
  });
  test(`005. tokenize error`, async () => {
    try {
      tokenize(`select * from "error`);
    } catch (err) {
      logger.log(err);
      expect(err).toMatchSnapshot('error');
    }
  });
  test(`006. tokenize error`, async () => {
    try {
      tokenize(`
    select /* from A`);
    } catch (err) {
      logger.log(err);
      expect(err).toMatchSnapshot('error');
    }
  });
});

