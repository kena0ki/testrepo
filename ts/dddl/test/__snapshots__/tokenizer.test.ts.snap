// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenize tokenize create statement: Create table 1`] = `
Array [
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "create",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "table",
  },
  Whitespace {
    "_value": " ",
  },
  DelimitedIdent {
    "_content": "ITEM",
    "_delimiter": "\\"",
    "_value": "\\"ITEM\\"",
  },
  Whitespace {
    "_value": " ",
  },
  LParen {
    "_value": "(",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "id",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "char",
  },
  LParen {
    "_value": "(",
  },
  Num {
    "_value": "10",
  },
  RParen {
    "_value": ")",
  },
  Comma {
    "_value": ",",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "price",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "decimal",
  },
  LParen {
    "_value": "(",
  },
  Num {
    "_value": "10",
  },
  Comma {
    "_value": ",",
  },
  Num {
    "_value": "3",
  },
  RParen {
    "_value": ")",
  },
  Comma {
    "_value": ",",
  },
  Whitespace {
    "_value": " ",
  },
  Cmmnt {
    "_value": "/* yen */",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Cmmnt {
    "_value": "-- qty int,",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  RParen {
    "_value": ")",
  },
  Operator {
    "_value": ";",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
]
`;

exports[`tokenize tokenize operators: query 1`] = `
Array [
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Operator {
    "_value": "\\\\",
  },
  Operator {
    "_value": "[",
  },
  Operator {
    "_value": "]",
  },
  Operator {
    "_value": "&",
  },
  Operator {
    "_value": "^",
  },
  Operator {
    "_value": "{",
  },
  Operator {
    "_value": "}",
  },
  Operator {
    "_value": "~",
  },
  Word {
    "_value": "#",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
]
`;

exports[`tokenize tokenize select statement: query 1`] = `
Array [
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  NewLine {
    "_value": "
",
  },
  Word {
    "_value": "SELECT",
  },
  NewLine {
    "_value": "
",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  HexStringLiteral {
    "_content": "'hex string'",
    "_prefix": "X",
    "_value": "X'hex string'",
  },
  Whitespace {
    "_value": " ",
  },
  Operator {
    "_value": "||",
  },
  Whitespace {
    "_value": " ",
  },
  HexStringLiteral {
    "_content": "'hex string2'",
    "_prefix": "x",
    "_value": "x'hex string2'",
  },
  Comma {
    "_value": ",",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  NationalStringLiteral {
    "_content": "'national string'",
    "_prefix": "N",
    "_value": "N'national string'",
  },
  Comma {
    "_value": ",",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  SingleQuotedString {
    "_value": "'character '' string'",
  },
  Comma {
    "_value": ",",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Num {
    "_value": "1",
  },
  Operator {
    "_value": "+",
  },
  Num {
    "_value": "2",
  },
  Operator {
    "_value": "-",
  },
  Num {
    "_value": "3",
  },
  Operator {
    "_value": "/",
  },
  Num {
    "_value": "4",
  },
  Operator {
    "_value": "*",
  },
  Num {
    "_value": "5",
  },
  Operator {
    "_value": "%",
  },
  Num {
    "_value": "6",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "from",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "A@B",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "C",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "D",
  },
  Comma {
    "_value": ",",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "_E",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "E",
  },
  Comma {
    "_value": ",",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "#F",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "where",
  },
  Whitespace {
    "_value": " ",
  },
  Operator {
    "_value": "!",
  },
  LParen {
    "_value": "(",
  },
  Word {
    "_value": "D",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "d",
  },
  Whitespace {
    "_value": " ",
  },
  Operator {
    "_value": "=",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "E",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "e",
  },
  RParen {
    "_value": ")",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "and",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "D",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "d",
  },
  Whitespace {
    "_value": " ",
  },
  Operator {
    "_value": "!=",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "E",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "e",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "and",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "D",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "d",
  },
  Whitespace {
    "_value": " ",
  },
  Operator {
    "_value": "<>",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "E",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "e",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "and",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "D",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "d",
  },
  Whitespace {
    "_value": " ",
  },
  Operator {
    "_value": "<",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "E",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "e",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "and",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "D",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "d",
  },
  Whitespace {
    "_value": " ",
  },
  Operator {
    "_value": ">",
  },
  Whitespace {
    "_value": " ",
  },
  Word {
    "_value": "E",
  },
  Operator {
    "_value": ".",
  },
  Word {
    "_value": "e",
  },
  Operator {
    "_value": ";",
  },
  NewLine {
    "_value": "
",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
  Whitespace {
    "_value": " ",
  },
]
`;
